chat-server/src/services/llm.ts

```diff
@@ -31,91 +31,85 @@ export type OpenAiStreamingResponse = AsyncIterable<
 
-export class OpenAiLlmProvider
-  implements LlmProvider<OpenAiStreamingResponse, OpenAiAwaitedResponse>
-{
-  private openAiChatClient: OpenAiChatClient;
+interface MakeOpenAiLlmParams {
+  apiKey: string;
+  deployment: string;
+  baseUrl: string;
+}
 
-  constructor(openAiChatClient: OpenAiChatClient) {
-    this.openAiChatClient = openAiChatClient;
-  }
+export function makeOpenAiLlm({
+  apiKey,
+  deployment,
+  baseUrl,
+}: MakeOpenAiLlmParams): Llm<OpenAiStreamingResponse, OpenAiAwaitedResponse> {
+  const openAiChatClient = new OpenAiChatClient(baseUrl, deployment, apiKey);
+  return {
+    // NOTE: for streaming implementation, see // NOTE: for example streaming data, see https://github.com/openai/openai-node/issues/18#issuecomment-1369996933
+    async answerQuestionStream({
+      messages,
+      chunks,
+    }: LlmAnswerQuestionParams): Promise<OpenAiStreamingResponse> {
+      const messagesForLlm = prepConversationForOpenAiLlm({ messages, chunks });
+      const completionStream = await openAiChatClient.chatStream({
+        messages: messagesForLlm,
+        options: { ...OPENAI_LLM_CONFIG_OPTIONS, stream: true },
+      });
+      return completionStream;
+    },
+    async answerQuestionAwaited({
+      messages,
+      chunks,
+    }: LlmAnswerQuestionParams): Promise<OpenAiChatMessage> {
+      const messagesForLlm = prepConversationForOpenAiLlm({ messages, chunks });
+      const {
+        choices: [choice],
+      } = await openAiChatClient.chatAwaited({
+        messages: messagesForLlm,
+        options: OPENAI_LLM_CONFIG_OPTIONS,
+      });
+      const { message } = choice;
+      if (!message) {
+        throw new Error("No message returned from OpenAI");
+      }
+      return message as OpenAiChatMessage;
+    },
+  };
+}
 
-  // NOTE: for streaming implementation, see // NOTE: for example streaming data, see https://github.com/openai/openai-node/issues/18#issuecomment-1369996933
-  async answerQuestionStream({
-    messages,
+function prepConversationForOpenAiLlm({
+  messages,
+  chunks,
+}: LlmAnswerQuestionParams): OpenAiChatMessage[] {
+  validateOpenAiConversation(messages);
+  const lastMessage = messages[messages.length - 1];
+  const newestMessageForLlm = GENERATE_USER_PROMPT({
+    question: lastMessage.content,
     chunks,
-  }: LlmAnswerQuestionParams): Promise<OpenAiStreamingResponse> {
-    const messagesForLlm = this.prepConversationForLlm({ messages, chunks });
-    const completionStream = await this.openAiChatClient.chatStream({
-      messages: messagesForLlm,
-      options: { ...OPENAI_LLM_CONFIG_OPTIONS, stream: true },
-    });
-    return completionStream;
+  });
+  return [...messages.slice(0, -1), newestMessageForLlm];
+}
+// TODO: consider adding additional validation that messages follow the pattern
+// system, assistant, user, assistant, user, etc.
+// Are there any other things which we should validate here?
+function validateOpenAiConversation(messages: OpenAiChatMessage[]) {
```

Comment:
I'm not sure I understand the purpose of this function. How would a `OpenAiChatMessage[]` get into an invalid state? That should inform whether to throw exceptions or use assertions. If exceptions are thrown due to user error, how would these errors be handled?

Reply:
dont recall the original intent here (it was originally a method of the class, so there's not any actual new logic)

thinking about it now: exceptions would most likely be thrown due to incorrectly written code elsewhere in the application, as opposed to user error. 

think it's worth keeping this here to make clear at this module's level what the shape of the conversation must be.